### ISSUE TITLE
Base.float mess up hessian

### ISSUE THREAD

**Issue by Non-Contradiction** [JuliaDiff/ReverseDiff.jl#107]
```
function ff(x)
    float(sum(x.*x))
end

using ReverseDiff

ReverseDiff.gradient(ff, [1.,2.])
# 2-element Array{Float64,1}: 
# 2.0 
# 4.0 

ReverseDiff.hessian(ff, [1.,2.])
# 2×2 Array{Float64,2}:
# 0.0  0.0
# 0.0  0.0
```

The `gradient` result is as expected while the `hessian` result is not.
I understand that the `float` function here is totally unnecessary.

**Comment by Non-Contradiction** [comment]
After investigation, the issue appears because in https://github.com/JuliaDiff/ReverseDiff.jl/blob/beb031ff6ae669eebcedbb4f9e380e8b41adf7bd/src/tracked.jl#L355
only `TrackedReal` with `V` as `AbstractFloat` get unchanged, but if you have a `TrackedReal` with `TrackedReal{V<:AbstractFloat}`, the problem will happen.  This is also why the issue occurs in `hessian` but not `gradient`. For example, the following thing will fix the issue:
```
julia> using ReverseDiff

julia> function ff(x) float(sum(x.^2)) end
ff (generic function with 1 method)

julia> ReverseDiff.hessian(ff, [1.,2.])
2×2 Array{Float64,2}:
 0.0  0.0
 0.0  0.0

julia> Base.float{V<:AbstractFloat,D,O}(t::ReverseDiff.TrackedReal{ReverseDiff.TrackedReal{V,D,O}}) = t

julia> ReverseDiff.hessian(ff, [1.,2.])
2×2 Array{Float64,2}:
 2.0  0.0
 0.0  2.0
```

**Comment by Non-Contradiction** [comment]
Looking into the code more closely. Before the aforementioned "fix(?)", the deriv information will be lost in
https://github.com/JuliaDiff/ReverseDiff.jl/blob/beb031ff6ae669eebcedbb4f9e380e8b41adf7bd/src/tracked.jl#L350
maybe it is better to preserve deriv information in the function above when "V <: TrackedReal

**Comment by jrevels** [comment]
That method seems like a premature optimization - if you simply remove it, does that fix your case?

**Comment by Non-Contradiction** [comment]
No. Removing it will further cause the `gradient` to break down, since the method at line 350 will not preserve the derivative information, and will assign `zeros` to the `D` field.

**Comment by Non-Contradiction** [comment]
It seems to me that the idea seems to be that 
(line 355) if the value is of `AbstractFloat` type, then reserve the derivative information, and
(line 350) if value is anything else, then discard the derivative information.
As I understand, this should work if we just want to get the first order derivative, but when we want to use it recursively, as in the case of `hessian` it will break down...
